### PR TITLE
Parsing: fetch <affected><resolution><component> from oval2 xml

### PR DESCRIPTION
**Pull request by daynewlee**

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -8,4 +8,5 @@ golang.org/x/text v0.3.0/go.mod h1:NqM8EUOU14njkJ3fqMW+pc6Ldnwhi/IjpwHt7yyuwOQ=
 golang.org/x/tools v0.0.0-20190924052046-3ac2a5bbd98a h1:DJzZ1GRmbjp7ihxzAN6UTVpVMi6k4CXZEr7A3wi2kRA=
 golang.org/x/tools v0.0.0-20190924052046-3ac2a5bbd98a/go.mod h1:b+2E5dAYhXwXZwtnZ6UAqBI28+e2cm9otk0dWdXHAEo=
 golang.org/x/xerrors v0.0.0-20190717185122-a985d3407aa7/go.mod h1:I/5z698sn9Ka8TeJc9MKroUUfqBBauWjQqLJ2OPfmY0=
+golang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543 h1:E7g+9GITq07hpfrRu66IVDexMakfv52eLZ2CXBWiKr4=
```
Comment (RTann) [inline go.sum]: how come this was updated?

File: oval/types.go [hunk 2]
```diff
@@ -89,16 +89,26 @@ type Reference struct {
...
+	Severity        string      `xml:"severity"`
+	Cves            []Cve       `xml:"cve"`
+	Bugzillas       []Bugzilla  `xml:"bugzilla"`
+	AffectedCPEList []string    `xml:"affected_cpe_list>cpe"`
+	Refs            []Ref       `xml:"ref"` // Ubuntu Only
+	Bugs            []Bug       `xml:"bug"` // Ubuntu Only
+	PublicDate      Date        `xml:"public_date"`
+	Issued          Date        `xml:"issued"`
+	Updated         Date        `xml:"updated"`
+	Affected        []Affected2 `xml:"affected"`
```
Comment (RTann) [inline oval/types.go]: This is probably Red Hat, only, right? Let's check, and if that's the case, let's add a comment similar to the comments on other fields `// Red Hat Only`

File: oval/types.go [hunk 3]
```diff
@@ -89,16 +89,26 @@ type Reference struct {
...
+}
+
+type Affected2 struct {
```
Comment (RTann) [inline oval/types.go]: Let's use a different name for this. Maybe `AffectedStatus`?

File: testdata/RHEL-8-including-unpatched-test.xml [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+</definitions>
+<variables>
+<local_variable comment="Get saved_entry in /boot/grub2/grubenv" datatype="int" id="oval:com.redhat.rhsa:var:20191167001" version="643">
+ <arithmetic arithmetic_operation="add">
+  <literal_component datatype="int">1</literal_component>
+  <object_component item_field="text" object_ref="oval:com.redhat.rhsa:obj:20191167027"/>
+ </arithmetic>
+</local_variable>
+</variables>
+</oval_definitions>
```
Comment (RTann) [inline testdata/RHEL-8-including-unpatched-test.xml]: nit: new line after this

File: go.sum (hunk 1 above)
Comment (daynewlee) [reply go.sum]: I really have no idea. But I will check out the original one.

File: oval/types.go [hunk 5]
```diff
@@ -89,16 +89,26 @@ type Reference struct {
...
+	Severity        string           `xml:"severity"`
+	Cves            []Cve            `xml:"cve"`
+	Bugzillas       []Bugzilla       `xml:"bugzilla"`
+	AffectedCPEList []string         `xml:"affected_cpe_list>cpe"`
+	Refs            []Ref            `xml:"ref"` // Ubuntu Only
+	Bugs            []Bug            `xml:"bug"` // Ubuntu Only
+	PublicDate      Date             `xml:"public_date"`
+	Issued          Date             `xml:"issued"`
+	Updated         Date             `xml:"updated"`
+	Affected        []AffectedStatus `xml:"affected"` // Red Hat Only
```
Comment (hdonnay) [inline oval/types.go]: Can there be multiple `affected` objects? If so, how are they differentiated?

File: oval/types.go [hunk 6]
```diff
@@ -89,16 +89,26 @@ type Reference struct {
...
+}
+
+type AffectedStatus struct {
+	Resolution []Resolution `xml:"resolution"`
```
Comment (hdonnay) [inline oval/types.go]: Can there be multiple `resolution` objects?

**Review by hdonnay — CHANGES_REQUESTED** [review]
Is there any documentation for this additional set of fields? It seems like it'd be very hard to use without documentation on things like "what does `component` contain" and "what are valid `state` attributes".

For concrete changes, the commits need to be flattened and rewritten to 
```
pkg: title

summary
```
format.

**Comment by daynewlee** [comment]
>

**Comment by daynewlee** [comment]
> Is there any documentation for this additional set of fields? It seems like it'd be very hard to use without documentation on things like "what does `component` contain" and "what are valid `state` attributes".
> 
> For concrete changes, the commits need to be flattened and rewritten to
> 
> ```
> pkg: title
> 
> summary
> ```
> 
> format.

The component means packages that's been affected by the current vulnerability. And state means the current condition of this vulnerability impacting the following packages, such as "Will not fix", "Affected", etc. I didn't find a formal documentation on those and I think those information are RHEL oval v2 only. 

Overall those information are informing people granular status for reported unpatched CVEs @hdonnay

**Review by hdonnay — CHANGES_REQUESTED** [review]
> The component means packages that's been affected by the current vulnerability. And state means the current condition of this vulnerability impacting the following packages, such as "Will not fix", "Affected", etc. I didn't find a formal documentation on those and I think those information are RHEL oval v2 only.
>
>Overall those information are informing people granular status for reported unpatched CVEs

My point is more that I shouldn't have to go look at the test fixures or the PR to figure out what these fields are expected to contain; please document that in the code.

The commits still need to be merged and the commit message rewritten into the normal form.

**Comment by daynewlee** [comment]
> 

Added new comments in the code with the documentation link
Updated the commit message

**Comment by daynewlee** [comment]
> > The component means packages that's been affected by the current vulnerability. And state means the current condition of this vulnerability impacting the following packages, such as "Will not fix", "Affected", etc. I didn't find a formal documentation on those and I think those information are RHEL oval v2 only.
> > Overall those information are informing people granular status for reported unpatched CVEs
> 
> My point is more that I shouldn't have to go look at the test fixures or the PR to figure out what these fields are expected to contain; please document that in the code.
> 
> The commits still need to be merged and the commit message rewritten into the normal form.


Added new comments in the code with the documentation link @hdonnay 
Updated the commit message

File: oval/types.go [hunk 7]
```diff
@@ -89,16 +89,30 @@ type Reference struct {
...
+// https://access.redhat.com/articles/5554431 section Oct 13, 2021 Oval
+type AffectedStatus struct {
+	Resolution []Resolution `xml:"resolution"`
+}
+
+// Resolution state values are one of:
+// Affected, Under investigation, Will not fix, Out of support scope, and Fix deferred
+type Resolution struct {
+	State     string   `xml:"state,attr"`
+	Component []string `xml:"component"`
```
Comment (RTann) [inline oval/types.go]: nit: `Components`

File: oval/types.go (hunk 6 above)
Comment (RTann) [reply oval/types.go]: nit: `Resolutions`

@hdonnay, yes, for example CVE-2019-25067:

```
<affected>
    <resolution state="Affected">
     <component>container-tools:rhel8/podman</component>
     <component>container-tools:rhel8/podman-catatonit</component>
     <component>container-tools:rhel8/podman-debugsource</component>
     <component>container-tools:rhel8/podman-docker</component>
     <component>container-tools:rhel8/podman-gvproxy</component>
     <component>container-tools:rhel8/podman-manpages</component>
     <component>container-tools:rhel8/podman-plugins</component>
     <component>container-tools:rhel8/podman-remote</component>
     <component>container-tools:rhel8/podman-tests</component>
    </resolution>
    <resolution state="Out of support scope">
     <component>container-tools:2.0/podman</component>
     <component>container-tools:2.0/podman-catatonit</component>
     <component>container-tools:2.0/podman-debugsource</component>
     <component>container-tools:2.0/podman-docker</component>
     <component>container-tools:2.0/podman-gvproxy</component>
     <component>container-tools:2.0/podman-manpages</component>
     <component>container-tools:2.0/podman-plugins</component>
     <component>container-tools:2.0/podman-remote</component>
     <component>container-tools:2.0/podman-tests</component>
    </resolution>
   </affected>
```

File: oval/resolution_test.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
+package oval
+
+import (
+	"encoding/xml"
+	"github.com/google/go-cmp/cmp"
+	"os"
+	"testing"
```
Comment (RTann) [inline oval/resolution_test.go]:
```suggestion
	"encoding/xml"
	"os"
	"testing"

	"github.com/google/go-cmp/cmp"
```

**Review by RTann — APPROVED** [review]
just one nit otherwise LGTM once Hank approves

File: oval/resolution_test.go (hunk 8 above)
Comment (hdonnay) [reply oval/resolution_test.go]: Yeah, I think this PR is good after this nit.

**Review by hdonnay — APPROVED** [review]

**Pull request merged**